LogicLayer/Friends/FriendsService.cs

```diff
@@ -0,0 +1,133 @@
+                .Union(friendsThatInvitedMe)
+                .OrderBy(u => u.UserName) // maybe order by NormalizedUserName? However, I cannot do that with UserDto.
+                .ToListAsync();
+
+            return allFriends;
+        }
+
+        public async Task SendFriendRequestAsync(CreateFriendRequestDto friendRequest)
+        {
+            if (friendRequest.UserNameTo == friendRequest.UserNameFrom)
+            {
+                throw new ApplicationException("You cannot invite yourself.");
+            }
+
+            var requestedById = await _repository.GetUserIdByUsername(friendRequest.UserNameFrom);
+            var requestedToId = await _repository.GetUserIdByUsername(friendRequest.UserNameTo);
+
+            _repository.Friendships.Add(
+                new Friendship()
+                {
+                    RequestedById = requestedById,
+                    RequestedToId = requestedToId,
+                    FriendshipStatus = FriendshipStatus.PendingFriendRequest
+                });
+
+            try
+            {
+                await _repository.SaveChangesAsync();
+            }
+            catch (DbUpdateException e) when (e.InnerException is SqlException sqlException)
+            {
+                switch (sqlException.Number)
+                {
+                    case 50000:
+                        throw new FriendRequestAlreadyPendingException(
+                            "You have already been invited by this user. Check your inbox for the friend request.");
+                    case 50001:
+                        throw new FriendRequestAlreadyPendingException(
+                            "You have already sent this user a friend request.");
+                    case 50002:
+                        throw new FriendRequestAlreadyFinalizedException("You are already friends.");
+                    default:
+                        throw new ArgumentOutOfRangeException();
+                }
+            }
+
+            //send email with link
+        }
+
+        public Task<FriendRequestDto> GetFriendRequestAsync(string friendRequestId)
+        {
+            var id = int.Parse(friendRequestId);
+            var friendship = _repository.Friendships.Where(f => f.Id == id);
+
+            return friendship
+                .Include(f => f.RequestedBy)
+                .Include(f => f.RequestedTo)
+                .Select(
+                    f => new FriendRequestDto(
+                        friendRequestId,
+                        new UserDto(f.RequestedBy.Id, f.RequestedBy.UserName, f.RequestedBy.Email),
+                        new UserDto(f.RequestedTo.Id, f.RequestedTo.UserName, f.RequestedTo.Email),
+                        f.FriendshipStatus.ToFriendRequest()))
+                .SingleOrDefaultAsync();
+        }
+
+        private async Task ChangeFriendRequestStatus(string friendRequestId, FriendshipStatus friendshipStatus)
+        {
+            var id = int.Parse(friendRequestId);
```

Comment:
Same as above, no reason for the parameter to be string anymore.

Reply:
Did it.